packages/vm/src/cache.rs

```diff
@@ -100,17 +111,32 @@ where
         deps: Extern<S, A, Q>,
         options: InstanceOptions,
     ) -> VmResult<Instance<S, A, Q>> {
-        // try from the module cache
-        let module = self.fs_cache.load(checksum)?;
-        if let Some(module) = module {
+        // Get module from memory cache
+        if let Some(module) = self.memory_cache.load(checksum)? {
+            self.stats.hits_memory_cache += 1;
+            let instance =
+                Instance::from_module(module, deps, options.gas_limit, options.print_debug)?;
+            return Ok(instance);
+        }
+
+        // Get module from file system cache
+        if let Some(module) = self.fs_cache.load(checksum)? {
             self.stats.hits_fs_cache += 1;
-            return Instance::from_module(&module, deps, options.gas_limit, options.print_debug);
+            let instance =
+                Instance::from_module(&module, deps, options.gas_limit, options.print_debug)?;
+            self.memory_cache.store(checksum, module)?;
```

Comment:
This compiles? Instance doesn't hold the reference to module?

Reply:
Jupp, the instance only needs a reference to the module for instantiation. It is cloned internally (but only a ref, not a deep-copy):

```rust
impl Module {
    // ...

    pub fn instantiate(&self, import_object: &ImportObject) -> error::Result<Instance> {
        Instance::new(Arc::clone(&self.inner), import_object)
    }
```